Readme.txt
General
-------

This extension allows you to secure files with apache2's mod_rewrite based on several criteria. The advantage of this solution is that the file paths stay the same, there is no PHP involved when serving the files (speed increase) and the headers are sent as if it were a normal apache request. For this the apache2-module mod_rewrite MUST be installed!
The criteria you can use to define whether a file is accessible are:
- is the file in a secured folder?
	-> In the standard .htaccess file it is assumed that all files inside fileadmin/, uploads/ and typo3temp/ should be secured!
- is the file in a public folder within the secured folders?
- is the user logged in as a BE_USER?
	-> By default BE_USERS may see all files. Anything else would be beyond the scope of this extension!
- is the file linked for the logged in user?
	-> For example within a <img>- or <a>-tag.
- is the file linked for any users not logged in?
	-> For example within a <img>- or <a>-tag.

Installation
------------

When installing this extensions, several things have to be done:
- complete and save the extension configuration
	-> Some important global values are stored there. Make sure you save it at least once (even if only default values are used)
- you must add the .htaccess section (described below) to your main .htaccess
- you must create at least one "delete" scheduler task
	-> this will clean up all the generated files
- you may create a "public indexer" scheduler task
	-> this is not mandatory, however the tx_securefiles_domain_model_public data records will not be considered in that case!

Scheduler Tasks
---------------




.htaccess
---------
In the .htaccess the main work is done! Unauthorized access is prohibited, so in order to have the extension do anything you must include the following lines.
You may change some of them where indicated, for example if your TYPO3 working directory is not the DOCUMENT_ROOT of the domain. In that case you may add the "RewriteBase /some/path" to indicate to apache2 the real directory TYPO3 is in. However please refer to the apache2 manual for more information on how to use RewriteBase.
As an alternative, you may inject the extra path in the conditions below.
You may also change the file suffixes that are excluded (per default .css and .js) and the folders that are secured by default.
However the RewriteEngine must be on!


#### Rewrite settings
RewriteEngine On

# start of secure_files section
# -----------------------------
#
# store the cookies in environment variables
#
# check if a BE_USER is logged in and get the hash
RewriteCond %{HTTP_COOKIE} tx_securefiles_be=([^;]+) [NC]
RewriteRule . - [env=tx_securefiles_be:%1]

# check if a FE_USER is logged in and get the hash
RewriteCond %{HTTP_COOKIE} tx_securefiles_fe=([^;]+) [NC]
RewriteRule . - [env=tx_securefiles_fe:%1]

# resolve the relative path
RewriteCond %{REQUEST_URI} ^([^\?]+)(\?.*)?$ [NC]
RewriteRule . - [env=tx_securefiles_rel:%1]

# realurl files don't exist and nc_staticfilecache files are not in the root folder
# we only care about the actually existing files
RewriteCond %{REQUEST_FILENAME} -f

# we only want to secure fileadmin/uploads/typo3temp
# all other folders should be typo3-standard stuff!
# should the TYPO3 not be situated in the root, the following rule has to be adapted
#   e.g. www.example.com/some/folder/<TYPO3 folder structure starts here> leads to
#   RewriteCond %{REQUEST_URI} ^/some/folder/(fileadmin|uploads|typo3temp)/
RewriteCond %{REQUEST_URI} ^/(fileadmin|uploads|typo3temp)/

# exclude any kind of files/folders next
# js/css files are usually public anyway
RewriteCond %{REQUEST_FILENAME} !\.(js|css)$

# and then check if the be user has global permission
# again, you might have to insert the /some/folder if you use a different root
RewriteCond %{DOCUMENT_ROOT}/typo3temp/tx_securefiles/be/%{ENV:tx_securefiles_be} !-f

# and then check if the fe user has the permission to the file
# again, you might have to insert the /some/folder if you use a different root
RewriteCond %{DOCUMENT_ROOT}/typo3temp/tx_securefiles/fe%{ENV:tx_securefiles_rel}/%{ENV:tx_securefiles_fe} !-f

# and then check if the file is public
# again, you might have to insert the /some/folder if you use a different root
RewriteCond %{DOCUMENT_ROOT}/typo3temp/tx_securefiles/pub%{ENV:tx_securefiles_rel} !-f

# lastly send forbidden in case neither be_user, fe_user permission or public file was detected!
RewriteRule . - [F]
#
# ---------------------------
# end of secure_files section


ChangeLog
2012-03-08 Nils Blattner <nb@cabag.ch>
	* version 0.0.8
	* changed the parsing of internal absolute links (http://this.domain.com/some/file.pdf) to use preg_match instead of parse_url (special characters do not work in parse_url)

2012-03-07 Nils Blattner <nb@cabag.ch>
	* version 0.0.7
	* fixed a bug with special characters in the urls

2012-03-05 Nils Blattner <nb@cabag.ch>
	* version 0.0.6
	* fixed a bug in the base url regex
	* added <link-tags to the base regex

2012-02-06 Nils Blattner <nb@cabag.ch>
	* version 0.0.5
	* fixed some bugs with the scheduler task

2012-02-06 Nils Blattner <nb@cabag.ch>
	* version 0.0.4
	* changed tx_SecureFiles_Utility_Parser to force create a .htaccess in the permission folder

2012-01-24 Nils Blattner <nb@cabag.ch>
	* version 0.0.3
	* changed tx_SecureFiles_Utility_Parser to only consider files

2011-12-21 Nils Blattner <nb@cabag.ch>
	* version 0.0.2
	* removed unnecessary function from the tx_SecureFiles_Utility_Parser class
	* added Tx_SecureFiles_Utility_Configuration class, that supplies the extension configuration with default values

2011-12-01 Nils Blattner <nb@cabag.ch>
	* version 0.0.1
	* initial release
	* added Readme.txt
	* added Tx_SecureFiles_Domain_Model_Public table/model/repository
	* added Scheduler task to recursively delete the auto generated files
	* added Scheduler task to create public entries based on tx_securefiles_domain_model_public entries


ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "secure_files".
#
# Auto generated 08-03-2012 11:20
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Secured files using mod_rewrite',
	'description' => 'This is an extension, that helps secure any files in the TYPO3-System.
The idea behind it is that:
- file paths stay the same
- configuration is easy
- folders can be marked public',
	'category' => 'misc',
	'author' => 'Nils Blattner',
	'author_email' => 'nb@cabag.ch',
	'author_company' => 'cab services ag',
	'shy' => '',
	'priority' => '',
	'module' => '',
	'state' => 'stable',
	'internal' => '',
	'uploadfolder' => '',
	'createDirs' => 'typo3temp/tx_securefiles/',
	'modify_tables' => '',
	'clearCacheOnLoad' => 0,
	'lockType' => '',
	'version' => '0.0.8',
	'constraints' => array(
		'depends' => array(
			'cms' => '',
			'typo3' => '4.3.0-0.0.0',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:28:{s:9:"ChangeLog";s:4:"1020";s:21:"ExtensionBuilder.json";s:4:"fc41";s:10:"Readme.txt";s:4:"d1d9";s:16:"ext_autoload.php";s:4:"de6d";s:21:"ext_conf_template.txt";s:4:"5d86";s:12:"ext_icon.gif";s:4:"e922";s:17:"ext_localconf.php";s:4:"3898";s:14:"ext_tables.php";s:4:"7890";s:14:"ext_tables.sql";s:4:"4272";s:31:"Classes/Domain/Model/Public.php";s:4:"0749";s:46:"Classes/Domain/Repository/PublicRepository.php";s:4:"f513";s:23:"Classes/Hooks/Login.php";s:4:"50d7";s:30:"Classes/Hooks/ParseContent.php";s:4:"baec";s:46:"Classes/Scheduler/AdditionalFieldsProvider.php";s:4:"438c";s:53:"Classes/Scheduler/CleanupAdditionalFieldsProvider.php";s:4:"eca1";s:33:"Classes/Scheduler/CleanupTask.php";s:4:"197f";s:33:"Classes/Utility/Configuration.php";s:4:"117d";s:26:"Classes/Utility/Parser.php";s:4:"8869";s:44:"Configuration/ExtensionBuilder/settings.yaml";s:4:"073f";s:28:"Configuration/TCA/Public.php";s:4:"cf5d";s:40:"Resources/Private/Language/locallang.xml";s:4:"0ac6";s:79:"Resources/Private/Language/locallang_csh_tx_securefiles_domain_model_public.xml";s:4:"e8db";s:43:"Resources/Private/Language/locallang_db.xml";s:4:"1317";s:35:"Resources/Public/Icons/relation.gif";s:4:"e615";s:61:"Resources/Public/Icons/tx_securefiles_domain_model_public.gif";s:4:"905a";s:46:"Tests/Unit/Controller/PublicControllerTest.php";s:4:"346a";s:38:"Tests/Unit/Domain/Model/PublicTest.php";s:4:"7e37";s:14:"doc/manual.sxw";s:4:"8d2d";}',
);

?>